### ISSUE TITLE
About the usage of SearchAbout and SearchPattern

### ISSUE THREAD

**Issue by jaewooklee93** [snu-sf/pl2015#71]
For proving `pigeonhole_principle`, I used the theorem `S n < S m -> n < m`.

However, the problem is that I don't know how I can find that theorem using `SearchPattern`. My first try was `SearchPattern (S _ < S _ -> _ < _)`, however, it didn't give any useful information. It only showed

```
Coq < SearchPattern (S _ < S _ -> _ < _).

Warning: query commands should not be inserted in scripts
```

So I tried again with `SearchPattern (S _ <= S _ -> _ <= _)`. At this time, I got the following result.

```
Coq < SearchPattern (S _ <= S _ -> _ <= _).

Warning: query commands should not be inserted in scripts
le_S_n: forall n m : nat, S n <= S m -> n <= m
Le.le_S_n: forall n m : nat, S n <= S m -> n <= m
Sn_le_Sm__n_le_m: forall n m : nat, S n <= S m -> n <= m
```

Thus, I just guessed that I could use `Lt.lt_S_n` for my purpose, since `Le.le_S_n` gave such result. That guess was found to be correct. After I had known the name if theorem, I could find its content with using `SearchAbout`.

```
Coq < SearchAbout "lt_S_n".

Warning: query commands should not be inserted in scripts
Lt.lt_S_n: forall n m : nat, S n < S m -> n < m
```

On the other hand, when I used `SearchAbout` instead of `SearchPattern` to look for the theorem related to `<=`, it gave 

```
Coq < SearchAbout le.

Warning: query commands should not be inserted in scripts
Between.between_le:
  forall (P : nat -> Prop) (k l : nat), Between.between P k l -> k <= l
Between.between_Sk_l:
  forall (P : nat -> Prop) (k l : nat),
  Between.between P k l -> S k <= l -> Between.between P (S k) l
Between.between_restr:
  forall (P : nat -> Prop) (k l m : nat),
  k <= l -> l <= m -> Between.between P k m ->
...

(1000 lines of incomprehensible search results)
```

So I think that `SearchAbout` is almost useless if I deal with built-in objects like `plus`, `mult` or `le`, since it brings us bunch of theorems we don't need as well. Is it right?

In conclusion, my questions are, 
1. In both of homework and exams, can we use all theorems which can be found by `SearchAbout` or `SearchPattern`, such as `le_S_n`, `Le.le_S_n` and `Sn_le_Sm__n_le_m`. Or, is there the most preferable one among them?
2. What is the proper way to find the theorem `S n < S m -> n < m` using `SearchPattern` in one-shot?
3. Can we use `SearchAbout` to search for theorems related to built-in objects? Since `SearchPattern` needs the exact pattern of theorem, it is sometimes harder to use than `SearchAbout`.

**Comment by jaewooklee93** [comment]
My question 1 is related to this phenomenon.

I'm curious about what happens if I have proved `Sn_le_Sm__n_le_m` of `Assignment05_32.v` with using `le_S_n`, which is exactly the same proposition. However, it doesn't work and show the error message.

``` Coq
(* Assignment05_32.v *)
Theorem Sn_le_Sm__n_le_m : forall n m,
  S n <= S m -> n <= m.
Proof. 
  apply le_S_n.
Qed.
```

```
Error: Impossible to unify "forall n m : nat, S n <= S m -> n <= m" with
 "forall n m : nat, S n <= S m -> n <= m".
```

However, when I have copied that proof into `Assignment06_08.v`, it works. Thus, I have two different cases, one which I can use `le_S_n` and another which I cannot use `le_S_n`, but I don't know the difference.

**Comment by fortunist** [comment]
2.
I found the way using <code>SearchAbout (Peano.lt (S _) (S _) -> Peano.lt _ _).</code>
Though I cannot explain why it does not work without <code>Require Export "Prop".</code>......

Addition :
It was because that we cannot use <code>Search, SearchAbout, SearchPattern</code> properly without the code <code>Require String. Open Scope string_scope.</code> in <code>induction.v</code>.

By the way, I think the reason why <code>SearchAbout lt.</code> cannot find <code>Lt.lt_S_n</code> is that the definition of <code>lt</code> was redefined in <code>Prop.v</code>. <code>SearchPattern (S _ < S _ -> _ < _).</code> find <code>Lt.lt_S_n</code> well without redefining <code>lt</code>.

The library <code>Lt</code> use the definition of <code>lt</code> in <code>Init/Peano.v</code>.Coq  cannot distinguish our <code>lt</code> and <code>Init/Peano.v</code>'s <code>lt</code> when searching. However, with the equivalence of both definitions, we can apply <code>Lt.lt_S_n</code> to the new definition, I presume.

**Comment by jaewooklee93** [comment]
@fortunist Oh, it really works. Thank you.
I guess the reason for that is `Prop.v` indirectly includes `Induction.v` and `Induction.v` imports lots of functions from the standard library.
By the way, since `SearchAbout lt` doesn't work but `SearchAbout Peano.lt` works, the problem may be due to the overlapped namespaces or scopes.

**Comment by fortunist** [comment]
Oh, you commented when I was editing my comment for addition...
Thank you.

**Comment by jaewooklee93** [comment]
@fortunist We arrived at the same conclusion :)
Then it seems that it's better to avoid using `lt`. Well, we can always substitute `lt` with `le`, using `unfold lt`.

**Comment by fortunist** [comment]
1. (related phenomenon)
The new defintion of <code>le</code> in <code>Assignment05_00.v</code> and the other of <code>Init/Peano.v</code> are not equiavalent. Assignment's <code>le</code> has type <code>nat -> nat -> Prop</code>, while initial <code>le</code> has type nat -> Prop. It seems the difference makes restriction on applying initial theorems in <code>Assignment05_*.v</code>.

3.
<code>SearchAbout</code> is truly more powerful than <code>SearchPattern</code>. Just change a word <code>SearchPattern</code> to <code>SearchAbout</code>, and more results will appear.

Now that I've written, it seems you already know these now...

**Comment by fortunist** [comment]
1 - related phenomenon.
The one defintion of <code>le</code> in <code>Assignment05_00.v</code> and the other of <code>Init/Peano.v</code> are not equiavalent. Results of <code>Print le.</code> and <code>Print Peano.le.</code> have a little difference. It seems the difference makes restriction on applying initial theorems in <code>Assignment05_*.v</code>.

3.
<code>SearchAbout</code> is truly more powerful than <code>SearchPattern</code>. Just change a word <code>SearchPattern</code> to <code>SearchAbout</code>, and more results will appear. <code>SearchAbout</code> find the following all patterns in environmental definitions' argument types and return types. For example, compare the results of <code>SearchAbout (_ <= _)</code> and <code>SearchPattern(_ <= _)</code>.

Now that I've written, it seems you already know these now... Anyway I'm commenting to confirm my considerations.

**Comment by jaewooklee93** [comment]
@fortunist 
1: Yeah, that seems the reason for all problems. Also, I want to point out that even if we have exactly the same definition for both of `le` and `Peano.le`, we cannot use the previous theorems. Maybe Coq has hidden serial numbers for differentiating each definition for the same name.

3: I only read the tutorial from Coq official website, so I didn't know that I also could use `SearchAbout` to match patterns, instead of `SearchPattern`. Thank you for letting me know. However, I think the fact that `SearchAbout` is giving us more results, doesn't directly mean that `SearchAbout` is more powerful. Because it gives too many results for `le` that even we cannot read them all.

When I compared the results of `SearchAbout (_ <= _)` and `SearchPattern (_ <= _)`, I found that `SearchAbout` gave every theorem which contains `(_ <= _)`, whereas `SearchPattern` only shows the theorems, whose **conclusion** contain `(_ <= _)`. so seaching with `SearchPattern` is somewhat more selective and I think `SeachPattern` can be useful to find exactly what we want.
